components/salsa-2022/src/function/specify.rs

```diff
@@ -100,6 +100,8 @@ where
         C::Key: TrackedStructInDb<DynDb<'db, C>>,
     {
         self.specify(db, key, value, |_| QueryOrigin::Field);
+        let database_key_index = self.database_key_index(key);
+        db.runtime().add_output(database_key_index.into());
```

Comment:
It may be more complicated than we previous thought.  What I'm doing here is adding the fields (`FunctionIngredient`) of tracked structs to the query output, so that we can mark all fields as valid (update `validate_at`) when we `deep_verify_memo` and find there are no changes.  

But there is still a problem when we try to read directly or indirectly the fields of the tracked struct created in the current query.
like 
https://github.com/salsa-rs/salsa/blob/f0fe691d3f902fab4ea3f419ea4f7493b46a5030/salsa-2022-tests/tests/tracked_fn_read_own_entity.rs#L29-L35
https://github.com/salsa-rs/salsa/blob/f0fe691d3f902fab4ea3f419ea4f7493b46a5030/salsa-2022-tests/tests/deletion-cascade.rs#L51-L60

Reply:
I was thinking if we can store the query (`DatabaseKeyIndex`) that created the field, when we find that a field may be outdated, we can check if the query stack contains the query that created the field, if yes, the old value is valid, otherwise, panic .

Reply:
Oh, right, I remembered that I spent a while tinkering with this setup. =) I think I had excluded them as outputs for efficiency reasons, but logically they *are* outputs. Let me take a look, thanks for giving this a try.

Reply:
Yes, it panics if we try to reuse `intermidiate_result` after creating a distinct input to trigger a new revision.
https://github.com/salsa-rs/salsa/blob/f0fe691d3f902fab4ea3f419ea4f7493b46a5030/salsa-2022-tests/tests/tracked_fn_read_own_entity.rs#L85-L107
```
thread 'red_herring' panicked at 'accessing fields of tracked struct "tracked_fn_read_own_entity::MyTracked" from older revisions'
```

Reply:
> I would expect that we would update the revision to the current revision in the call to `specify`?

We would update the revision to the current revision in the current query when we *execute* it, but not when we *resue (deep verify)* it.

Reply:
OK, sorry, what I meant @XFFXFF was -- now that you've added these as outputs -- are those tests failing.

Reply:
What we should probably do is just remove the `specify_field` method...

https://github.com/salsa-rs/salsa/blob/f8f6dbd3490a439d970fde4020e06bbe7432b60f/components/salsa-2022/src/function/specify.rs#L98-L103

...and only have `specify_and_record`...

https://github.com/salsa-rs/salsa/blob/f8f6dbd3490a439d970fde4020e06bbe7432b60f/components/salsa-2022/src/function/specify.rs#L105-L109

...`specify_field` was a micro-optimization and I think it was wrong. I think we can make tracked structs more efficient, but for now this makes sense until we get everything right.

Reply:
This probably means we can remove `QueryOrigin::Field` altogether.

Reply:
I see one test still fails, I don't 100% know why, but I can dig into it later or tomorrow if you get stuck, @XFFXFF .

Reply:
> now that you've added these as outputs -- are those tests failing

now only [tracked_fn_read_own_entity::red_herring](https://github.com/salsa-rs/salsa/blob/f0fe691d3f902fab4ea3f419ea4f7493b46a5030/salsa-2022-tests/tests/tracked_fn_read_own_entity.rs#L85-L107) and [deletion-cascade](https://github.com/salsa-rs/salsa/blob/f0fe691d3f902fab4ea3f419ea4f7493b46a5030/salsa-2022-tests/tests/deletion-cascade.rs#L90-L130) panic. They all panic with something like   
```
thread 'red_herring' panicked at 'accessing fields of tracked struct "tracked_fn_read_own_entity::MyTracked" from older revisions'
```
The panic is what we add in this pr. Let me try to explain why it panics.  


Take the example as 
```rust
 #[salsa::tracked(jar = Jar)] 
 fn intermediate_result(db: &dyn Db, input: MyInput) -> MyTracked { 
     db.push_log(format!("intermediate_result({:?})", input)); 
     let tracked = MyTracked::new(db, input.field(db) / 2); 
     let _ = tracked.field(db); // read the field of an entity we created 
     tracked 
 } 

fn main() {
    ...
    let input = MyInput::new(&mut db, 22)
    intermediate_result(&db, input)  // OK

    let input2 = MyInput::new(&mut db, 33)
    intermediate_result(&db, input) // panic
}
```

We create an input and call `intermediate_result`, it's ok. Now we create a different input to trigger a new revision, and then call `intermediate_result` with original input (which hasn't changed), it panics. 

The second time we call `intermediate_reuslt(&db, input)`, we need to deep verify the result because of the new revision introduced by `input2`, and need to check the dependencies of the query. Since we call `tracked.field(db)` in `intermediate_result`, so `tracked.field(db)` is one of the dependencies of the query, we should verify `tracked.field(db)` first, then we will find the `verified_at` of `tracked.field(db)`  smaller than the current revision, so panic. 

We will mark `tracked.field(&db)` as valid (update `verified_at` to the current revision) after we deep verify `intermediate_result`, and then it is OK to use `tracked.field(&db)`, but we use it before the deep verification is done

```
intermediate_result -> deep verify intermediate_reuslt -> deep verify tracked.field -> panic
```

Reply:
OK, I see. This seems like a pre-existing bug that would affect specified, as well. Basically we validate all inputs, but we don't consider that some inputs are also outputs? 

I think first off we should add a test case like...

```rust
#[tracked] struct Test { }

#[tracked(specify)] fn specifiable(db: &dyn crate::Db, test: Test) -> u32 { 0 }

#[tracked] fn test(db: &dyn crate::Db) -> u32 {
    let t = Test::new();
    specifiable::specify(db, t, 22);
    specifiable(db, t)
}
```

If I understand the problem, this `test` function would also panic once an input is changed.

Then, I think the best fix would be to replace this loop

https://github.com/salsa-rs/salsa/blob/32f0ce0cffc0e3695a3292fb740d378cdcf19408/components/salsa-2022/src/function/maybe_changed_after.rs#L190-L195


with something like this

```rust
                let last_verified_at = old_memo.verified_at.load();
                for &input in edges.inputs().iter() {
                    if !edges.has_output(input) { // <-- NEW!
                        if db.maybe_changed_after(input, last_verified_at) {
                            return false;
                        }
                    }
                }
```

where `edges.has_output` is some new method we add to `QueryEdges`. It looks like we'll have to add back a helper method `outputs` and then we can use [`binary_search`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) to search through it (the outputs array is sorted).

Reply:
Alternatively, you're right that we could remember the query that specified the value of the field, and then modify `maybe_changed_after` so that it takes as input the id of the query being re-validated as well. 

What I don't like about it is that it makes the interface of that function more complex -- i.e., now it is saying "return true if the value have changed after this revision *or* it was assigned by this query". 

What I *do* like about it is that we don't have to search the outputs for *every* input.  There are sometimes quite a lot of inputs/outputs, so that could be a major change.

Reply:
OK, I did a quick search to look at who calls `maybe_changed_after` and it seems like the only real caller is this loop right here. I think @XFFXFF it's probably better to follow your suggestion, and modify `QueryOrigin::Specified` to include the query that specified the value. Then we can add a parameter to `maybe_changed_after` with a name like `consumer` or something, with the expected value being something like "has it changed after this revision OR was it created by this consumer". 

All things considered, I suspect that my suggestion of doing an extra binary search is going to be too expensive.